experimental/builder/include/ck_tile/builder/testing/conv_fwd.hpp

```diff
@@ -0,0 +1,259 @@
+#include "ck_tile/builder/testing/tensor_buffer.hpp"
+#include "ck/library/utility/convolution_parameter.hpp"
+#include "ck/library/utility/convolution_host_tensor_descriptor_helper.hpp"
+
+/// This file implements common functionality for invoking/testing grouped
+/// forward convolutions created through the CK Builder API. The main item
+/// of it is the ConvArgs structure - which contains a complete description
+/// of a convolution operation.
+///
+/// It is not intended that this file contains implementation details for
+/// actually launching a convolution operation. As this can be done
+/// through different APIs depending on the kernel (CK, CK Tile, or a
+/// reference implementation), the code dealing with that is split out
+/// into a separate header for each implementation.
+
+namespace ck_tile::builder::test {
+
+/// @brief Convolution tensor dimensions.
+///
+/// This structure is used to describe lengths of a convolution problem. In
+/// fact, this structure is a complete description of ALL inputs and outputs
+/// lengths of a convolution problem, as this structure contains all of the
+/// combined parameters. Note that we can't also use this structure to describe
+/// tensor strides: whereas the lengths are all governed by a common set of
+/// parameters, strides of the input, weight, and output tensor are all
+/// independent.
+template <int SPATIAL_DIM>
+struct ConvTensorLengths
+{
+    size_t batch_size          = 1;  // N
+    size_t groups              = 1;  // G
+    size_t input_channels      = 1;  // C
+    size_t output_channels     = 1;  // K
+    Extent<SPATIAL_DIM> image  = {}; // W, H, D
+    Extent<SPATIAL_DIM> filter = {}; // X, Y, Z
+};
+
+/// @brief `Args` specialization for forward convolution.
+///
+/// @tparam SIGNATURE Forward convolution signature.
+///
+/// @see Args
+template <auto SIGNATURE>
+    requires ValidConvSignature<SIGNATURE> && ConvDirectionIsForward<SIGNATURE>
+struct Args<SIGNATURE>
+{
+    constexpr static auto SPATIAL_DIM = SIGNATURE.spatial_dim;
+    constexpr static auto INPUT_TYPE  = SIGNATURE.data_type;
+    constexpr static auto WEIGHT_TYPE = SIGNATURE.data_type;
+    constexpr static auto OUTPUT_TYPE = SIGNATURE.data_type;
+
+    using Ops = factory::internal::ElementwiseOps<get_elementwise_operation<SIGNATURE>()>;
+
+    ConvTensorLengths<SPATIAL_DIM> lengths;
+    // TODO: Tensor strides. This needs a new structure as well as some
+    // reworking of the make_*_descriptor() functions, as the current
+    // implementation (based on ConvParam in old CK / CK Tile) does not
+    // support strides at all.
+
+    Extent<SPATIAL_DIM> filter_strides;
+    Extent<SPATIAL_DIM> filter_dilation;
+    Extent<SPATIAL_DIM> input_left_pad;
+    Extent<SPATIAL_DIM> input_right_pad;
+
+    Ops::AElementwiseOp a_elementwise_op;
+    Ops::BElementwiseOp b_elementwise_op;
+    Ops::CDEElementwiseOp cde_elementwise_op;
+
+    // TODO: We shouldn't need to call into an internal namespace here.
```

Comment:
This is an interesting problem. The internal namespace of the factory is private implementation translate the signature and algorithm to instantiate CK kernels. In principle it should not be used here, and this reuse is hinting that we may need some more general functionality for processing signature descriptions. I like your approach of doing the reuse and using a TODO, this helps us learn and keeps us from over-designing too early.